src/Integration.Vsix/SonarAnalyzerDeactivationManager.cs

```diff
@@ -0,0 +1,137 @@
+﻿//-----------------------------------------------------------------------
+// <copyright file="SonarAnalyzerDeactivationManager.cs" company="SonarSource SA and Microsoft Corporation">
+//   Copyright (c) SonarSource SA and Microsoft Corporation.  All rights reserved.
+//   Licensed under the MIT License. See License.txt in the project root for license information.
+// </copyright>
+//-----------------------------------------------------------------------
+
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.VisualStudio.ComponentModelHost;
+using Microsoft.VisualStudio.LanguageServices;
+using SonarLint.Helpers;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using SonarLint.VisualStudio.Integration.SonarAnalyzer;
+
+namespace SonarLint.VisualStudio.Integration.Vsix
+{
+    public class SonarAnalyzerDeactivationManager : IDisposable
+    {
+        internal /*for testing purposes*/ enum ProjectAnalyzerStatus
+        {
+            NoAnalyzer,
+            SameVersion,
+            DifferentVersion
+        }
+
+        private readonly Workspace workspace;
+        private readonly IActiveSolutionBoundTracker activeSolutionBoundTracker;
+
+        private static readonly AssemblyName AnalyzerAssemblyName =
+            new AssemblyName(typeof(SonarAnalysisContext).Assembly.FullName);
+        internal /*for testing purposes*/ static readonly Version AnalyzerVersion = AnalyzerAssemblyName.Version;
+        internal /*for testing purposes*/ static readonly string AnalyzerName = AnalyzerAssemblyName.Name;
+
+        internal /*for testing purposes*/ SonarAnalyzerDeactivationManager(IServiceProvider serviceProvider, Workspace workspace)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            this.workspace = workspace;
+            this.activeSolutionBoundTracker = serviceProvider.GetMefService<IActiveSolutionBoundTracker>();
+
+            SonarAnalysisContext.ShouldAnalysisBeDisabled =
+                tree => ShouldAnalysisBeDisabledOnTree(tree);
+        }
+
+        public SonarAnalyzerDeactivationManager(IServiceProvider serviceProvider)
```

Comment:
Consider refactoring the code so that one constructor will call a different constructor. Constructing the instance in two different completely ways (i.e. different code paths) is not a good practice. The common code code be in a private initialize method.